### ISSUE TITLE
🐛 error when using `createPlatformPayPaymentMethod` for Apple Pay on physical devices

### ISSUE THREAD

**Issue by walkerdb** [stripe/stripe-react-native#1236]
The following code returns a working token in development but in production builds it returns an error saying "The payment has been cancelled":

```ts
const cartItems = [{
  paymentType: PlatformPay.PaymentType.Immediate,
  label: 'Company Name',
  amount: '0.01',
  isPending: true,
}];

const { token, error } = await createPlatformPayPaymentMethod({
  applePay: { merchantCountryCode: 'US', currencyCode: 'USD', cartItems },
});
```

Visually on iOS calling this opens up the apple pay card selection form, then a success sound plays after authenticating with TouchID, then the form closes itself, but the function itself returns an `error` object whose `message` field is the above error message.

Are we doing something wrong here? As far as we can tell we've set things up correctly with our merchant ID and publishable key in the `StripeProvider`, and the exact same values were working when using tipsi-stripe. Would love pointers here if you have any.

Test Devices:
* iPhone SE 2nd gen, iOS 15.0.1
* iPhone 13 Pro, iOS 16.0

**Comment by walkerdb** [comment]
Update: in some cases we're seeing the message "This transaction approval has already been used to create a token." instead

**Comment by charliecruzan-stripe** [comment]
Is it possible that you're calling `createPlatformPayPaymentMethod` twice?

**Comment by walkerdb** [comment]
@charliecruzan-stripe according to all of our logs we're only calling it once

**Comment by charliecruzan-stripe** [comment]
Interesting, well it seems from the message that we're getting [here](https://github.com/stripe/stripe-react-native/blob/master/ios/ApplePayViewController.swift#L42), but at that time `applePaymentMethodFlowCanBeCanceled` _should_ be false (bc iOS triggers [this function first](https://github.com/stripe/stripe-react-native/blob/master/ios/ApplePayViewController.swift#L17)), and the only time it's ever set to true is [here at the beginning of the `createPlatformPayPaymentMethod` method](https://github.com/stripe/stripe-react-native/blob/master/ios/StripeSdk.swift#L557), which is why I ask if there's any chance it's getting triggered twice.

**Comment by walkerdb** [comment]
@charliecruzan-stripe 

More specifically what seems to be happening is that the first time a user tries to set up apple pay on a physical device with this code, they get

```
This transaction approval has already been used to create a token.
```

then each successive time they try to set up the card afterwards they get

```
The payment has been cancelled
```

I have no idea where the first message is coming from - it doesn't seem to be from code in this repo or in `stripe-ios`, nor does it show up anywhere in google except for a link to this issue. I'm assuming then that it must be a stripe backend API response?

Also wondering given the fact that future calls hit the `The payment has been cancelled` code path if that particular failure mode exits the swift code before `applePaymentMethodFlowCanBeCanceled` is ever set to false again after it was initially set to `true` the first time.

**Comment by walkerdb** [comment]
Looking at our Stripe logs, it seems that after completing the apple pay form summoned by `createPlatformPayPaymentMethod`, `stripe-react-native` is making 3 API calls:

1. A call to the `token` endpoint (response: 200)
2. A call to the `payment_methods` endpoint, using the token from the the first call (response: 200)
3. A _second_ call to the `token` endpoint (response: 400)

The second token call is what is failing and triggering the error logic. Based on the error message this may be because both token calls are using the same `pk_token_transaction_id`?

Some request details:

1. `/v1/tokens`
    ```json
    // POST request body
    {
      "pk_token_instrument_name": "Visa 7418",
      "pk_token_payment_network": "PKPaymentNetwork(_rawValue: Visa)",
      "pk_token": "******",
      "payment_user_agent": "stripe-ios/23.3.0; variant.legacy",
      "pk_token_transaction_id": "C3D876BB540E1AD8EB012826A538E2B7D1D3785966E9446B2FF0C4155DC27EBA"
    }
    ```
    ```json
    // response body
    {
      "id": ["tok_1MDeLHCtqB39Ngsoil2SMsAX"],
      "object": "token",
      "card": {
        "id": "card_1MDeLHCtqB39NgsoucCXfJ4I",
        "object": "card",
        "address_city": null,
        "address_country": null,
        "address_line1": null,
        "address_line1_check": null,
        "address_line2": null,
        "address_state": null,
        "address_zip": null,
        "address_zip_check": null,
        "brand": "Visa",
        "country": "US",
        "cvc_check": null,
        "dynamic_last4": "3313",
        "exp_month": 12,
        "exp_year": 2023,
        "funding": "credit",
        "last4": "7418",
        "name": null,
        "tokenization_method": "apple_pay"
      },
      "client_ip": "136.24.100.170",
      "created": 1670722115,
      "livemode": true,
      "type": "card",
      "used": false
    }
    ```
2. `/v1/payment_methods`
    ```json
    // POST request body
    {
      "type": "card",
      "payment_user_agent": "stripe-ios/23.3.0; variant.legacy",
      "card": {
        "token": "tok_1MDeLHCtqB39Ngsoil2SMsAX"
      }
    }
    ```
    ```json
    // response body
    {
      "id": "pm_1MDeLHCtqB39NgsoJEpH8vVd",
      "object": "payment_method",
      "billing_details": {
        "address": {
          "city": null,
          "country": null,
          "line1": null,
          "line2": null,
          "postal_code": null,
          "state": null
        },
        "email": null,
        "name": null,
        "phone": null
      },
      "card": {
        "brand": "visa",
        "checks": {
          "address_line1_check": null,
          "address_postal_code_check": null,
          "cvc_check": null
        },
        "country": "US",
        "exp_month": 12,
        "exp_year": 2023,
        "funding": "credit",
        "generated_from": null,
        "last4": "7418",
        "networks": {
          "available": [
            "visa"
          ],
          "preferred": null
        },
        "three_d_secure_usage": {
          "supported": true
        },
        "wallet": {
          "apple_pay": {
          },
          "dynamic_last4": "3313",
          "type": "apple_pay"
        }
      },
      "created": 1670722115,
      "customer": null,
      "livemode": true,
      "type": "card"
    }
    ```
3. `/v1/tokens`
    ```json
    // POST request body
    {
      "pk_token_payment_network": "PKPaymentNetwork(_rawValue: Visa)",
      "pk_token": "******",
      "pk_token_transaction_id": "C3D876BB540E1AD8EB012826A538E2B7D1D3785966E9446B2FF0C4155DC27EBA",
      "payment_user_agent": "stripe-ios/23.3.0; variant.legacy",
      "muid": "74587f94-3cd7-4db4-ac54-0d376e02ebdce4fae0",
      "pk_token_instrument_name": "Visa 7418",
      "guid": "03cd52c4-50f7-45f7-87fb-8dec62975bd72a97bf",
      "sid": "3992de73-3f1a-4d00-bc16-381f7dcd54a7c87cad"
    }
    ```
    ```json
    {
      "error": {
        "message": "This transaction approval has already been used to create a token.",
        "request_log_url": "https://dashboard.stripe.com/logs/req_tqZALcJ0ZYryn7?t=1670722115",
        "type": "invalid_request_error"
      }
    }
    ```

**Comment by walkerdb** [comment]
[this line in stripe-ios](https://github.com/stripe/stripe-ios/blob/78a89114636a70bf5f34d9c8371d0a17127e70e3/StripePayments/StripePayments/API%20Bindings/STPAPIClient%2BApplePay.swift#L152) may be why it's working in sim just fine - it auto-gens a random `pk_token_transaction_id` every time in sim, but uses the `pk_token_transaction_id` on the PKPayment object otherwise. 

I don't really have much context here but it seems like this might be fixable by mutating the payment to add a new transaction ID just before calling the final createToken here at the start of the `else` clause? If `payment` is mutable at least. https://github.com/stripe/stripe-react-native/blob/9a1a1636d7477df2d0bd72c3513daca5a12c263b/ios/ApplePayViewController.swift#L18-L35

**Comment by charliecruzan-stripe** [comment]
the best long-term solution here may just be to move out the `token` creation to a separate method, and so instead we'd have:
- `createPlatformPayPaymentMethod`
- `createPlatformPayToken`

where each one accepts the exact same params, and migrating from the latter to the former is trivial. There doesn't seem to be a situation where one needs both the PaymentMethod & a Token.

**Comment by charliecruzan-stripe** [comment]
Fix was released in [v0.23.0](https://github.com/stripe/stripe-react-native/releases/tag/v0.23.0)
